sound/soc/sof/ipc4-topology.c

```diff
@@ -1650,12 +1655,31 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 		control_data = scontrol->ipc_control_data;
 		if (control_data->data->blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
 			continue;
-
+		data_size = control_data->data->size;
 		data = (void *)control_data->data->data;
-		memcpy(cfg, data, control_data->data->size);
+
 		break;
 	}
 
+	/* Make sure we have enough data for base module extension structure */
+	if (swidget->init_blob_style == SOF_INIT_BLOB_BASE_WITH_BASE_EXT) {
+		struct sof_ipc4_base_module_cfg_ext *base_ext;
+		int size;
+
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)data;
+		size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
+					base_ext->priv_param_len +
+					(base_ext->num_sink_pin_fmts +
+					base_ext->num_source_pin_fmts) *
+					sizeof(struct sof_ipc4_pin_format);
+		if (data_size != size) {
+			dev_err(sdev->dev, "Insufficient data for widget %s initialization\n",
+				swidget->widget->name);
```

Comment:
should this check be done on line 1660? What if there are no kcontrols for the widget?

Reply:
what if there are more than one kcontrols for the widget?